syntax = "proto3";

package xgboost.federated;

service Federation {
  rpc Allreduce(AllreduceRequest) returns (AllreduceReply) {}
}

enum DataType {
  CHAR = 0;
  UCHAR = 1;
  INT = 2;
  UINT = 3;
  LONG = 4;
  ULONG = 5;
  FLOAT = 6;
  DOUBLE = 7;
  LONGLONG = 8;
  ULONGLONG = 9;
}

enum ReduceOperation {
  MAX = 0;
  MIN = 1;
  SUM = 2;
}

message AllreduceRequest {
  bytes send_buffer = 1;
  DataType data_type = 2;
  ReduceOperation reduce_operation = 3;
}

message AllreduceReply {
  bytes receive_buffer = 1;
}
